src/pyfmi/fmi3.pyx

```diff
@@ -675,6 +678,155 @@ cdef class FMUModelBase3(FMI_BASE.ModelBase):
+                matching (filter="*der*"). Can also be a list of filters
+                See http://docs.python.org/2/library/fnmatch.html.
+                Default: None
+
+        Returns::
+
+            Dict with variable name as key and a pyfmi.fmi3.FMI3ModelVariable class as value.
+        """
+        cdef FMIL3.fmi3_import_variable_t*        variable
+        cdef FMIL3.fmi3_import_variable_list_t*   variable_list
+        cdef FMIL.size_t                          variable_list_size
+        cdef FMIL3.fmi3_value_reference_t         value_ref
+        cdef FMIL3.fmi3_base_type_enu_t           data_type
+        cdef FMIL3.fmi3_variability_enu_t         data_variability
+        cdef FMIL3.fmi3_causality_enu_t           data_causality
+        cdef FMIL3.fmi3_initial_enu_t             data_initial
+
+        cdef int has_start = 0
+        # TODO: Can we rename the keyword arg 'filter' to variable_filter?
+        variable_filter = filter
+        # TODO: Can we rename the keyword arg 'type' to variable_type?
+        variable_type = type
+        cdef list filter_list, variable_return_list = []
+        variable_dict = {}
+
+        variable_list      = FMIL3.fmi3_import_get_variable_list(self._fmu, 0)
+        variable_list_size = FMIL3.fmi3_import_get_variable_list_size(variable_list)
+
+        """
+            TODO: If we decide to force user to use FMI3_* enums as input we need to also error check
+                  such that user doesn't specify for example causality = FMI3_Type.X or similar.
+                  Also let's remove '_as_list'?
+        """
+
+        for index in range(variable_list_size):
+            variable = FMIL3.fmi3_import_get_variable(variable_list, index)
+
+            has_start = FMIL3.fmi3_import_get_variable_has_start(variable)
+            if only_start and has_start == 0:
+                continue
+
+            data_variability = FMIL3.fmi3_import_get_variable_variability(variable)
+
+            if only_fixed:
+                # fixed variability requires start-value
+                if has_start == 0:
+                    continue
+                elif data_variability != FMI3_Variability.FIXED.value:
+                    continue
+
+            name             = pyfmi_util.decode(FMIL3.fmi3_import_get_variable_name(variable))
+            value_ref        = FMIL3.fmi3_import_get_variable_vr(variable)
+            description      = self._get_variable_description(variable)
+            data_causality   = FMIL3.fmi3_import_get_variable_causality(variable)
+            data_initial     = FMIL3.fmi3_import_get_variable_initial(variable)
+            data_type        = FMIL3.fmi3_import_get_variable_base_type(variable)
+            #If only variables with start are wanted, check if the variable has start
+
+            # TODO: Discuss if we want to support also regular integers as inputs
+            if isinstance(variable_type, FMI3_Type) and (data_type != variable_type.value):
```

Comment:
Not sure if Cython will be smart enough to only evaluate these once? Might want to explicitly calculate it only once like we did in FMI2.

Reply:
I think this should be fine, it is also `get_model_variables` which is not used repeatedly during simulation and we had`if ... is not None` in the FMI2 specification so I think the difference is minimal if anything.

Reply:
See d8aaeea